document/core/exec/instructions.rst

```diff
@@ -254,23 +254,474 @@ Reference Instructions
+
+8. If the :ref:`reference type <syntax-reftype>` :math:`\X{rt}_2` :ref:`matches <match-reftype>` :math:`\X{rt}_1`, then:
+
+   a. Push the value :math:`\I32.\CONST~1` to the stack.
+
+9. Else:
+
+   a. Push the value :math:`\I32.\CONST~0` to the stack.
 
 .. math::
    \begin{array}{lcl@{\qquad}l}
-   \val~\REFASNONNULL &\stepto& \TRAP
-     & (\iff \val = \REFNULL~\X{ht}) \\
-   \val~\REFASNONNULL &\stepto& \val
+   S; \reff~(\REFTEST~\X{rt}) &\stepto& (\I32.\CONST~1)
+     & (\iff S \vdashval \reff : \X{rt}'
+        \land \vdashreftypematch \X{rt}' \matchesreftype \insttype_{F.\AMODULE}(\X{rt})) \\
+   S; \reff~(\REFTEST~\X{rt}) &\stepto& (\I32.\CONST~0)
      & (\otherwise) \\
    \end{array}
 
 
+.. _exec-ref.cast:
+
+:math:`\REFCAST~\X{rt}`
+.......................
+
+1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+
+2. Let :math:`\X{rt}_1` be the :ref:`reference type <syntax-reftype>` :math:`\insttype_{F.\AMODULE}(\X{rt})`.
+
+3. Assert: due to :ref:`validation <valid-ref.test>`, :math:`\X{rt}_1` is :ref:`closed <type-closed>`.
+
+4. Assert: due to :ref:`validation <valid-ref.test>`, a :ref:`reference value <syntax-ref>` is on the top of the stack.
+
+5. Pop the value :math:`\reff` from the stack.
+
+6. Assert: due to validation, the :ref:`reference value <syntax-ref>` is :ref:`valid <valid-ref>` with some :ref:`reference type <syntax-reftype>`.
+
+7. Let :math:`\X{rt}_2` be the :ref:`reference type <syntax-reftype>` of :math:`\reff`.
+
+8. If the :ref:`reference type <syntax-reftype>` :math:`\X{rt}_2` :ref:`matches <match-reftype>` :math:`\X{rt}_1`, then:
+
+   a. Push the value :math:`\reff` back to the stack.
+
+9. Else:
+
+   a. Trap.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   S; \reff~(\REFCAST~\X{rt}) &\stepto& \reff
+     & (\iff S \vdashval \reff : \X{rt}'
+        \land \vdashreftypematch \X{rt}' \matchesreftype \insttype_{F.\AMODULE}(\X{rt})) \\
+   S; \reff~(\REFCAST~\X{rt}) &\stepto& \TRAP
+     & (\otherwise) \\
+   \end{array}
+
+
+
+.. _exec-i31.new:
+
+:math:`\I31NEW`
+...............
+
+1. Assert: due to :ref:`validation <valid-extern.externalize>`, a :ref:`value <syntax-val>` of :ref:`type <syntax-valtype>` |I32| is on the top of the stack.
+
+2. Pop the value :math:`(\I32.\CONST~i)` from the stack.
+
+3. Let :math:`j` be the result of computing :math:`\wrap_{32,31}(i)`.
+
+4. Push the reference value :math:`(\REFI31~j)` to the stack.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   (\I32.\CONST~i)~\I31NEW &\stepto& (\REFI31~\wrap_{32,31}(i))
+   \end{array}
+
+
+.. _exec-i31.get_s:
+
+:math:`\I31GET\K{\_}\sx`
+........................
+
+1. Assert: due to :ref:`validation <valid-extern.externalize>`, a :ref:`value <syntax-val>` of :ref:`type <syntax-valtype>` :math:`(\REF~\NULL~\I31)` is on the top of the stack.
+
+2. Pop the value :math:`\reff` from the stack.
+
+3. If :math:`\reff` is :math:`\REFNULL~t`, then:
```

Comment:
We cannot change the type annotation, since it can originate from the source program. E.g., `(ref.null i31) (i31.get_s)` is of course a well-typed source program, so must be allowed in the reduction rule.

This is one of the warts caused by introducing type-annotated null, there are more (e.g., the typing rule for ref.null values in exec/values). That's why I never liked it. ;)

Reply:
I see, thanks. I guess an alternative would be to "decode" the source `ref.null t` to `ref.null none` at some point, but I can see that there are benefits to keeping the binary format and abstract syntax more in sync.